File: apps/frontend/lib/format.ts
import { dictionaries, type Lang } from "./i18n";

const compactViews = new Intl.NumberFormat("en-US", {
  notation: "compact",
  maximumFractionDigits: 1,
});

export function formatViewCount(viewCount: string, lang: Lang = "en"): string {
  const n = Number(viewCount);
  const suffix = dictionaries[lang].format.views;
  if (!Number.isFinite(n)) return `— ${suffix}`;
  return `${compactViews.format(n)} ${suffix}`;
}

export function formatRelativeTime(iso: string, lang: Lang = "en"): string {
  const date = new Date(iso);
  if (Number.isNaN(date.getTime())) return "";

  const t = dictionaries[lang].format;
  const diffMinutes = Math.round((Date.now() - date.getTime()) / 60_000);
  if (diffMinutes < 60) return t.minutesAgo(Math.max(diffMinutes, 0));

  const diffHours = Math.round(diffMinutes / 60);
  if (diffHours < 24) return t.hoursAgo(diffHours);

  const diffDays = Math.round(diffHours / 24);
  if (diffDays < 7) return t.daysAgo(diffDays);

  const diffWeeks = Math.round(diffDays / 7);
  return t.weeksAgo(diffWeeks);
}

File: apps/mobile/lib/utils/format.dart
import 'package:intl/intl.dart';

import '../i18n/dictionary.dart';
import '../i18n/lang.dart';

final NumberFormat _compactViews = NumberFormat.compact(locale: 'en_US');

String formatViewCount(String viewCount, AppLang lang) {
  final n = num.tryParse(viewCount);
  final suffix = dictionaries[lang]!.viewsSuffix;
  if (n == null) return '— $suffix';
  return '${_compactViews.format(n)} $suffix';
}

String formatRelativeTime(String iso, AppLang lang) {
  final date = DateTime.tryParse(iso);
  if (date == null) return '';

  final t = dictionaries[lang]!;
  final diffMinutes = DateTime.now().difference(date).inMinutes;
  if (diffMinutes < 60) return t.minutesAgo(diffMinutes < 0 ? 0 : diffMinutes);

  final diffHours = (diffMinutes / 60).round();
  if (diffHours < 24) return t.hoursAgo(diffHours);

  final diffDays = (diffHours / 24).round();
  if (diffDays < 7) return t.daysAgo(diffDays);

  final diffWeeks = (diffDays / 7).round();
  return t.weeksAgo(diffWeeks);
}
